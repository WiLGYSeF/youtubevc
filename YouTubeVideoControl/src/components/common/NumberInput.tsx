import Decimal from 'decimal.js-light';
import React, { ChangeEvent, KeyboardEvent, useState } from 'react';

import { clamp } from '../../utils/clamp';

interface NumberInputProps {
  label?: string;
  labelLeft?: boolean;

  minValue?: number;
  maxValue?: number;
  step?: number | null;
  defaultValue: number;

  clamp?: boolean;
  forceValue?: boolean;

  setValue: (value: number) => void;
}

function coerceNumber(
  value: number,
  minValue?: number,
  maxValue?: number,
  step?: number,
  doClamp?: boolean,
): number {
  let num = value;

  if (doClamp) {
    if (minValue !== undefined && maxValue === undefined) {
      num = Math.max(num, minValue);
    } else if (minValue === undefined && maxValue !== undefined) {
      num = Math.min(num, maxValue);
    } else if (minValue !== undefined && maxValue !== undefined) {
      num = clamp(num, minValue, maxValue);
    }
  } else if (minValue !== undefined && maxValue !== undefined) {
    const diff = maxValue - minValue;

    for (; num < minValue; num += diff);
    for (; num > maxValue; num -= diff);
  }

  if (step) {
    num = new Decimal(num).div(step).toInteger().mul(step)
      .toNumber();
  }

  return num;
}

function NumberInput(props: NumberInputProps) {
  const [number, setNumber] = useState(props.defaultValue);
  const [value, setValue] = useState(props.defaultValue.toString());

  const labelLeft = props.labelLeft ?? true;
  const doClamp = props.clamp ?? true;
  const forceValue = props.forceValue ?? false;

  let step: number | undefined;
  if (props.step === undefined) {
    step = 1;
  } else if (props.step === null) {
    step = undefined;
  } else {
    step = props.step;
  }

  const eLabel = (
    <span>{props.label}</span>
  );

  const doForceValue = () => {
    if (forceValue) {
      setValue(number.toString());
    }
  };

  if (doClamp
    && props.minValue === undefined
    && props.maxValue === undefined
  ) {
    throw new Error('cannot have undefined min and max values when clamping');
  }

  return (
    <label>
      {labelLeft && eLabel}
      <input
        type="number"
        min={props.minValue} max={props.maxValue} step={step ?? 'any'}
        value={value}
        onChange={(e: ChangeEvent<HTMLInputElement>) => {
          const num = coerceNumber(
            Number(e.target.value),
            props.minValue,
            props.maxValue,
            step,
            doClamp,
          );

          props.setValue(num);
          setNumber(num);
          setValue(e.target.value);
        }}
        onBlur={doForceValue}
        onKeyUp={(e: KeyboardEvent<HTMLInputElement>) => {
          if (e.key === 'Enter') {
            doForceValue();
          }
        }}
      />
      {!labelLeft && eLabel}
    </label>
  );
}

export default NumberInput;
