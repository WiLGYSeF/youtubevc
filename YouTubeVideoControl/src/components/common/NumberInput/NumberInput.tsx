import React, { ChangeEvent, KeyboardEvent } from 'react';

import coerceNumber from 'utils/coerceNumber';
import useStatePropBacked from 'utils/useStatePropBacked';

interface NumberInputProps {
  label?: string;
  labelRight?: boolean;

  minValue?: number;
  maxValue?: number;
  step?: number | null;
  value: number;

  clamp?: boolean;
  forceValue?: boolean;

  onChange(value: number): void;
}

const NUMBER_REGEX = /^-?\d*\.\d*$/;
const NONDIGIT_REGEX = /[^\d.-]/g;

function NumberInput(props: NumberInputProps) {
  const [number, setNumber] = useStatePropBacked(props.value);
  const [value, setValue] = useStatePropBacked(props.value.toString());

  const labelRight = props.labelRight ?? false;
  const doClamp = props.clamp ?? true;
  const forceValue = props.forceValue ?? false;

  let step: number | undefined;
  if (props.step === undefined) {
    step = 1;
  } else if (props.step === null) {
    step = undefined;
  } else {
    step = props.step;
  }

  const eLabel = (
    <span>{props.label}</span>
  );

  const doForceValue = () => {
    if (forceValue) {
      setValue(number.toString());
    }
  };

  return (
    <label data-testid="number-input" data-label-right={labelRight}>
      {!labelRight && eLabel}
      <input
        min={props.minValue} max={props.maxValue} step={step ?? 'any'}
        value={value}
        onChange={(e: ChangeEvent<HTMLInputElement>) => {
          const sanitized = e.target.value.replace(NONDIGIT_REGEX, '');
          const val = Number(sanitized);

          const num = coerceNumber(
            val,
            props.minValue,
            props.maxValue,
            step,
            doClamp,
          );

          props.onChange(num);
          setNumber(num);
          setValue(sanitized);
        }}
        onBlur={doForceValue}
        onKeyUp={(e: KeyboardEvent<HTMLInputElement>) => {
          if (e.key === 'Enter') {
            doForceValue();
          }
        }}
      />
      {labelRight && eLabel}
    </label>
  );
}

export default NumberInput;
