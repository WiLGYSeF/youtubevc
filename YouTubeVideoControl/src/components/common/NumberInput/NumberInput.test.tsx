import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import flatten from 'utils/flattenElements';
import NumberInput from './NumberInput';

export interface NumberInputInputs {
  input: HTMLInputElement;
}

export function getInputs(container: HTMLElement): NumberInputInputs {
  return {
    input: container.querySelector('input')!,
  };
}

function getLabel(container: HTMLElement): HTMLElement {
  return container.querySelector('[data-testid="label"]') as HTMLElement;
}

describe('NumberInput', () => {
  it('renders label', () => {
    const { container } = render(<NumberInput
      label="test label"
      defaultValue={0}
      onChange={() => {}}
    />);

    const label = getLabel(container);
    const { input } = getInputs(container);

    const flattened = flatten(container);

    expect(flattened.indexOf(label)).toBeLessThan(flattened.indexOf(input));
  });

  it('renders label on right', () => {
    const { container } = render(<NumberInput
      label="test label"
      labelRight
      defaultValue={0}
      onChange={() => {}}
    />);

    const label = getLabel(container);
    const { input } = getInputs(container);

    const flattened = flatten(container);

    expect(flattened.indexOf(input)).toBeLessThan(flattened.indexOf(label));
  });

  it.each([
    // basic tests
    ['123', undefined, undefined, null, true, false, [1, 12, 123], '123'],
    ['1.23', undefined, undefined, null, true, false, [1, 1, 1.2, 1.23], '1.23'],
    ['-37', undefined, undefined, null, true, false, [NaN, -3, -37], '-37'],
    ['1az', undefined, undefined, null, true, false, [1, 1, 1], '1'],
    ['1.2.3', undefined, undefined, null, true, false, [1, 1, 1.2, 1.2, 1.23], '1.23'],

    // clamp tests
    ['123', 100, 120, null, true, false, [100, 100, 120], '123'],
    ['123', 100, 120, null, true, true, [100, 100, 120], '120'],
    ['123', undefined, 120, null, true, false, [1, 12, 120], '123'],
    ['123', undefined, 120, null, true, true, [1, 12, 120], '120'],

    // wrap tests
    ['1000', 0, 360, null, false, false, [1, 10, 100, 280], '1000'],
    ['1000', 0, 360, null, false, true, [1, 10, 100, 280], '280'],

    // step tests
    ['1.27', undefined, undefined, 0.1, true, false, [1, 1, 1.2, 1.3], '1.27'],
    ['1.27', undefined, undefined, 0.1, true, true, [1, 1, 1.2, 1.3], '1.3'],
  ])(
    'inputs "%s", [%d, %d] +%d clamp %d forceValue %d',
    (
      inputStr: string,
      minValue: number | undefined,
      maxValue: number | undefined,
      step: number | null,
      clamp: boolean,
      forceValue: boolean,
      expected: number[],
      expectedStr: string,
    ) => {
      let testValue = 0;
      const testFn = jest.fn();

      const { container } = render(<NumberInput
        minValue={minValue}
        maxValue={maxValue}
        step={step}
        defaultValue={testValue}
        clamp={clamp}
        forceValue={forceValue}
        onChange={(value: number) => {
          testValue = value;
          testFn(value);
        }}
      />);

      const { input } = getInputs(container);

      expect(input.value).toEqual(testValue.toString());

      userEvent.clear(input);
      testFn.mockClear();

      for (let i = 0; i < inputStr.length; i += 1) {
        userEvent.keyboard(inputStr[i]);
        expect(testValue).toEqual(expected[i]);
      }

      fireEvent.blur(input);

      expect(input.value).toEqual(expectedStr);
      expect(testValue).toEqual(expected[expected.length - 1]);

      expect(testFn.mock.calls.map((x) => x[0])).toEqual(expected);
    },
  );

  it('forces update on enter', () => {
    const { container } = render(<NumberInput
      minValue={1}
      maxValue={10}
      defaultValue={15}
      clamp
      forceValue
      onChange={() => {}}
    />);

    const { input } = getInputs(container);

    expect(input.value).toEqual('15');

    userEvent.clear(input);
    userEvent.type(input, '3{enter}');
    expect(input.value).toEqual('3');

    userEvent.clear(input);
    userEvent.type(input, '15{enter}');
    expect(input.value).toEqual('10');
  });

  it('updates value on props change', () => {
    let value = 5;

    const { container, rerender } = render(<NumberInput
      minValue={1}
      maxValue={20}
      defaultValue={value}
      clamp
      forceValue
      onChange={() => { }}
    />);

    const { input } = getInputs(container);

    expect(input.value).toEqual(value.toString());

    value = 17;
    rerender(<NumberInput
      minValue={1}
      maxValue={20}
      defaultValue={value}
      clamp
      forceValue
      onChange={() => { }}
    />);

    expect(input.value).toEqual(value.toString());
  });
});
