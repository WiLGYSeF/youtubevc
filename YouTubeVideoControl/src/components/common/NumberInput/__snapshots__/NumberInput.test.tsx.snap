// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput renders label 1`] = `
<label
  data-label-right={false}
  data-testid="number-input"
>
  <span>
    test label
  </span>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    step={1}
    value="0"
  />
</label>
`;

exports[`NumberInput renders label on right 1`] = `
<label
  data-label-right={true}
  data-testid="number-input"
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    step={1}
    value="0"
  />
  <span>
    test label
  </span>
</label>
`;
